### PR TITLE
545 strategy backtests add financial metrics

### PR DESCRIPTION
**Pull request by gabrielfior**
-> Added Sharpe ratio, mean daily return and annual volatility as extra variables when selecting strategies

### CODE REVIEW

File: examples/monitor/financial_metrics.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        self.risk_free_rate = risk_free_rate
+
+    def __has_df_valid_columns_else_exception(self,
+                                              required_columns: list[str]
+                                              ) -> None:
+        if not set(required_columns).issubset(self.df.columns):
+            raise ValueError("Dataframe doesn't contain all the required columns.")
+
+    def prepare_wallet_daily_balance_df(self, timestamp_col_name: str,
+                                        profit_col_name: str) -> pd.DataFrame:
```
Comment (gabrielfior) [inline examples/monitor/financial_metrics.py:22]: Simply generate a df that entails the daily wallet balance
Day | balance (xDAI)
1 | 50
2 | 52
...

File: examples/monitor/financial_metrics.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        # We generate a new Dataframe with daily wallet balances, derived by the final wallet balance
+        # from the previous day.
+        wallet_balance_daily_df = df2[['profit_cumsum']].resample('D').ffill()
+        wallet_balance_daily_df.dropna(inplace=True)
+        wallet_balance_daily_df["returns"] = wallet_balance_daily_df["profit_cumsum"].pct_change()
+        return wallet_balance_daily_df
+
+    # ToDo - Write tests
+    def calculate_annual_sharpe_ratio(self, timestamp_col_name: str = "timestamp",
+                                      profit_col_name: str = "sim_profit") -> SharpeOutput:
```
Comment (gabrielfior) [inline examples/monitor/financial_metrics.py:42]: I added `profit_col_name` as arg to allow this to be used also for the original bets Sharpe calculation (column `bet_profit`), see usage https://github.com/gnosis/prediction-market-agent-tooling/pull/548/files#diff-8df61548bdb13e9d38216f5c88637ff50c852babd5d5a488bff2b5c43b7a1081R294

File: examples/monitor/data_models.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from pydantic import BaseModel
```
Comment (kongzii) [inline examples/monitor/data_models.py:1]: Shouldn't this be outside of examples folder?

File: ape-config.yaml [hunk 4]
```diff
@@ -8,7 +8,7 @@ default_ecosystem: gnosis
 node:
   gnosis:
     mainnet:
-      uri: https://rpc.gnosischain.com
+      uri: https://rpc.gnosis.gateway.fm #https://rpc.gnosischain.com
```
Comment (kongzii) [inline ape-config.yaml:11]: Should `https://rpc.gnosischain.com` be removed?

File: examples/monitor/financial_metrics.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+import pandas as pd
+
+from examples.monitor.data_models import SharpeOutput, SimulationDetail
+
+
+class SharpeRatioCalculator:
```
Comment (kongzii) [inline examples/monitor/financial_metrics.py:7]: Also this can be probably defined outside of examples folder?

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 6]
```diff
@@ -1,6 +1,20 @@
...
+from examples.monitor.financial_metrics import SharpeRatioCalculator
+from examples.monitor.transaction_cache import TransactionBlockCache
+from prediction_market_agent_tooling.markets.omen.omen_contracts import (
+    OmenConditionalTokenContract,
+)
+from prediction_market_agent_tooling.markets.omen.omen_subgraph_handler import (
+    OmenSubgraphHandler,
+)
+
+dotenv.load_dotenv()
```
Comment (kongzii) [inline examples/monitor/match_bets_with_langfuse_traces.py:17]: Why is this needed? Aren't APIKeys working as expected?

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 7]
```diff
@@ -72,18 +85,34 @@ def get_outcome_for_trace(
...
+            else -buy_trade.amount.amount
+        )
 
-    correct = buy_trade.outcome == market_outcome
-    profit = (
-        received_outcome_tokens - buy_trade.amount.amount
-        if correct
-        else -buy_trade.amount.amount
-    )
+    # received_outcome_tokens = market.get_buy_token_amount(
```
Comment (kongzii) [inline examples/monitor/match_bets_with_langfuse_traces.py:112]: Some commented code, can it be removed or added another comment why it's commented out?

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 8]
```diff
@@ -165,12 +194,17 @@ def get_outcome_for_trace(
...
+        web3=OmenConditionalTokenContract().get_web3()
+    )
+
     for agent_name, private_key in agent_pkey_map.items():
         print(f"\n## {agent_name}\n")
         api_keys = APIKeys(BET_FROM_PRIVATE_KEY=private_key)
 
         # Pick a time after pool token number is stored in OmenAgentMarket
-        start_time = utc_datetime(2024, 10, 1)
+        start_time = utc_datetime(2024, 10, 28)
```
Comment (kongzii) [inline examples/monitor/match_bets_with_langfuse_traces.py:207]: Why is this necessary?

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 9]
```diff
@@ -211,51 +245,68 @@ def get_outcome_for_trace(
...
-                        "agent_p_yes": round(trace.answer.p_yes, 4),
-                        "agent_conf": round(trace.answer.confidence, 4),
-                        "org_bet": round(bet.amount.amount, 4),
-                        "sim_bet": round(simulated_outcome.size, 4),
-                        "org_dir": bet.outcome,
-                        "sim_dir": simulated_outcome.direction,
-                        "org_profit": round(bet.profit.amount, 4),
-                        "sim_profit": round(simulated_outcome.profit, 4),
-                    }
+                simulation_detail = SimulationDetail(
```
Comment (kongzii) [inline examples/monitor/match_bets_with_langfuse_traces.py:280]: 👍

File: examples/monitor/transaction_cache.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import diskcache as dc
+import tenacity
+from eth_typing import HexStr
+from tenacity import wait_exponential
+from web3 import Web3
+
+from prediction_market_agent_tooling.loggers import logger
+
+
+class TransactionBlockCache:
```
Comment (kongzii) [inline examples/monitor/transaction_cache.py:10]: I think this file should also be outside of examples folder?

File: prediction_market_agent_tooling/markets/manifold/data_models.py [hunk 11]
```diff
@@ -200,7 +200,6 @@ def get_profit(self, market_outcome: bool) -> ProfitAmount:
             if self.get_resolved_boolean_outcome() == market_outcome
             else -self.amount
         )
-        profit -= self.fees.get_total()
```
Comment (kongzii) [inline prediction_market_agent_tooling/markets/manifold/data_models.py:203]: I remember Evan saying that he copied Manifold implementation into Omen, did you verify that fees on Manifold are automatically deducted as well?

Comment (kongzii) [reply prediction_market_agent_tooling/markets/manifold/data_models.py:203]: oh I only now noticed that in omen model it isn't removed, maybe you meant to remove it only in omen's get_profit function?

File: examples/monitor/match_bets_with_langfuse_traces.py (hunk 6 above)
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:17]: This is needed for 2 things
->  `GNOSIS_RPC_URL` defined in .env - the variable is loaded when PMAT is imported, should actually be done via APIKeys (agreed?)
-> Google_API_Credentials are also defined in .env - should this be ported over to APIKeys?

If I do these 2 things, `load_dotenv` can be removed.

File: examples/monitor/match_bets_with_langfuse_traces.py (hunk 8 above)
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:207]: I wanted to see less data, but no reason to commit. Removing.

File: examples/monitor/match_bets_with_langfuse_traces.py (hunk 7 above)
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:112]: Removed

File: examples/monitor/transaction_cache.py (hunk 10 above)
Comment (gabrielfior) [reply examples/monitor/transaction_cache.py:10]: Agreed - removed

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 12]
```diff
@@ -72,18 +87,29 @@ def get_outcome_for_trace(
...
-        received_outcome_tokens - buy_trade.amount.amount
-        if correct
-        else -buy_trade.amount.amount
-    )
+    # If not correct, stop early because profit is known.
+    if not correct:
+        profit = -buy_trade.amount.amount
+    else:
+        # We use a historical state (by passing in a block_number as arg) to get the correct outcome token balances.
+        tx_block_number = tx_block_cache.get_block_number(actual_placed_bet.id)
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:96]: `get_buy_token_amount` is block-dependent, since the price of the token is dictated by the size of the `yes_outcome_pool_size` and `no_outcome_pool_size`

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 13]
```diff
@@ -72,18 +87,29 @@ def get_outcome_for_trace(
...
+
+        received_outcome_tokens = omen_agent_market_at_block.get_buy_token_amount(
+            bet_amount=omen_agent_market_at_block.get_bet_amount(
+                buy_trade.amount.amount
+            ),
+            direction=buy_trade.outcome,
+        ).amount
+        profit = (
+            received_outcome_tokens - buy_trade.amount.amount
+            if correct
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:110]: Actually this was correct, but readability was horrible - please see updated version

File: examples/monitor/match_bets_with_langfuse_traces.py [hunk 14]
```diff
@@ -72,18 +87,29 @@ def get_outcome_for_trace(
...
+            else -buy_trade.amount.amount
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:111]: This is not needed - see refactoring I mentioned earlier

File: ape-config.yaml (hunk 4 above)
Comment (gabrielfior) [reply ape-config.yaml:11]: There were some instabilities with this RPC (see Slack channel), but now it should be solved - reverting back

File: examples/monitor/financial_metrics.py (hunk 5 above)
Comment (gabrielfior) [reply examples/monitor/financial_metrics.py:10]: Agreed, placed inside `monitor` folder

File: examples/monitor/match_bets_with_langfuse_traces.py (hunk 6 above)
Comment (gabrielfior) [reply examples/monitor/match_bets_with_langfuse_traces.py:16]: Cool, this thread is addressed in the other PR - hence closing this

**Review by kongzii — APPROVED** [review]
